### PR TITLE
[Don't merge yet] Authentication and Sign in System

### PR DESCRIPTION
**Pull request by fatclarence**
I think the APIs are quite clear. 

However I don't know how I can better implement it with React-Router-Dom. Most of the react files I edited include 

AppRouter.js
SignIn.js
Register.js
PetOwnerHome.js

The rest of my files can be found in the utils (JsonWebTokens generations and helpers for validation)
Also the notifications are done using react-toastify.

I have an error in PetOwnerHome.js, whereby the return statements does not wait till the data to return from the database to check whether the person is a petowner. 

I have instantiated ENUMS in the schema for ADMIN, OWNER and CARETAKER. Can take a look at schema.sql!
I have also added a proxy for development environment. 

NOTE: DO NOT MERGE, as I have switched out some codes in dbPool to facilitate testing with my own localdb. Do give me all feedback and I will make changes this Saturday!

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -13,14 +13,9 @@
       }
     },
     "@babel/compat-data": {
-      "version": "7.11.0",
```
Comment (seeziyang) [inline package-lock.json:16]: I think we can stick to using either npm or yarn only so there won't be 2 different package-lock/yarn.lock files

File: server/utils/jwtGenerator.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+const jwt = require("jsonwebtoken");
+require(`dotenv`).config();
+
+function jwtGenerator(user_id) {
+    const payload = {
+        userId: user_id
+    }
+    
+    return jwt.sign(payload, process.env.jwtSecret, { expiresIn: "1hr" });
```
Comment (seeziyang) [inline server/utils/jwtGenerator.js:9]: maybe can let it expire after a longer time period 1hr feels quite short LOL but just my personal pref

File: server/utils/authorization.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+const jwt = require("jsonwebtoken");
+require("dotenv").config();
```
Comment (seeziyang) [inline server/utils/authorization.js:2]: i think dont have to config dotenv again cos we did it once in server.js already

File: server/utils/jwtGenerator.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+const jwt = require("jsonwebtoken");
+require(`dotenv`).config();
```
Comment (seeziyang) [inline server/utils/jwtGenerator.js:2]: same as prev

File: src/routes/AppRouter.js [hunk 5]
```diff
@@ -11,16 +13,36 @@ import Register from "../pages/Register";
...
-      <PublicRoute path={ROUTES.REGISTER} component={Register} />
-
-      <PrivatePetOwnerRoute path={ROUTES.PET_OWNER_HOME} component={PetOwnerHome} />
-      <PublicRoute path={ROUTES.HOME} exact={true} component={Home} />
-    </Switch>
-  </BrowserRouter>
-);
+
+
+const AppRouter = () =>{
```
Comment (seeziyang) [inline src/routes/AppRouter.js:18]: i think might have to use React Context to store user data (username of logged in user) when authenticated

not v sure too

**Review by seeziyang — APPROVED** [review]

File: src/routes/AppRouter.js (hunk 5 above)
Comment (jon-chua) [reply src/routes/AppRouter.js:18]: Yeah I was thinking we can wrap the whole thing in an `authContext` and then each of the routes like `PrivatePetOwnerRoute` will implement the individual authentication checking

Comment (fatclarence) [reply src/routes/AppRouter.js:18]: Will look into this. Thanks guys!

**Pull request merged**